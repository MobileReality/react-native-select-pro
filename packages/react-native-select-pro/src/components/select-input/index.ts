export { SelectInput } from './select-input';
