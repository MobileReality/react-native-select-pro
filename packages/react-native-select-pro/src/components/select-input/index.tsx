import React, { useEffect, useRef } from 'react';
import {
    I18nManager,
    Keyboard,
    StyleSheet,
    TextInput,
    TextStyle,
    TextInputProps,
} from 'react-native';

import { COLORS, FONT_SIZE } from '../../constants/styles';
import type { OptionalToRequired } from '../../helpers/types/OptionalToRequired';
import { Action, DispatchType, State } from '../../state/types';
import type {
    OnPressSelectControlType,
    OnSetPosition,
    SelectProps,
} from '../../types';

type SelectInputProps = OptionalToRequired<
    Pick<State, 'isOpened' | 'searchValue' | 'selectedOption'> & {
        dispatch: DispatchType;
    } & Pick<
            SelectProps,
            | 'placeholderText'
            | 'searchPattern'
            | 'disabled'
            | 'multiSelection'
            | 'selectControlTextStyle'
            | 'placeholderTextColor'
        > & { onPressSelectControl: OnPressSelectControlType } & {
            setPosition: OnSetPosition;
        } & {inputProps?: TextInputProps}
>;

export const SelectInput = ({
    disabled,
    isOpened,
    searchValue,
    searchPattern,
    placeholderText,
    onPressSelectControl,
    dispatch,
    setPosition,
    multiSelection,
    selectedOption,
    placeholderTextColor,
    selectControlTextStyle,
    inputProps,
}: SelectInputProps) => {
    const searchInputRef = useRef<TextInput>(null);

    useEffect(() => {
        const showSubscription = Keyboard.addListener('keyboardDidShow', () => {
            setPosition();
        });
        const hideSubscription = Keyboard.addListener('keyboardDidHide', () => {
            setPosition();
        });
        dispatch({
            type: Action.SetSearchInputRef,
            payload: searchInputRef,
        });
        return () => {
            showSubscription.remove();
            hideSubscription.remove();
            dispatch({
                type: Action.SetSearchInputRef,
                payload: null,
            });
        };
    }, []);

    const onChangeText = (payload: string) => {
        if (!disabled) {
            if (!isOpened) {
                dispatch({ type: Action.Open });
            }
            dispatch({
                type: Action.SetSearchValue,
                payload,
            });
            if (searchPattern) {
                dispatch({
                    type: Action.SearchOptions,
                    searchPattern,
                    payload,
                });
            }
        }
    };

    const resolvePlaceholder = () => {
        if (
            multiSelection &&
            Array.isArray(selectedOption) &&
            selectedOption.length > 0
        ) {
            return '  ';
        }
        return placeholderText;
    };

    return (
        <TextInput
            ref={searchInputRef}
            accessibilityLabel="Place text"
            editable={!disabled}
            placeholder={resolvePlaceholder()}
            placeholderTextColor={placeholderTextColor}
            style={
                disabled
                    ? [
                          styles.disabled,
                          styles.text,
                          multiSelection && { marginRight: 5 },
                      ]
                    : [styles.text, selectControlTextStyle]
            }
            textAlign={I18nManager.isRTL ? 'right' : 'left'}
            value={searchValue}
            onChangeText={onChangeText}
            onPressIn={disabled ? () => null : onPressSelectControl}
            {...inputProps}
        />
    );
};

type Styles = {
    text: TextStyle;
    disabled: TextStyle;
};

const styles = StyleSheet.create<Styles>({
    text: {
        fontSize: FONT_SIZE,
    },
    disabled: {
        backgroundColor: COLORS.DISABLED,
    },
});
