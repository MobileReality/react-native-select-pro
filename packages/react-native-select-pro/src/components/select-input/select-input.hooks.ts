import { useEffect, useRef } from 'react';
import type { TextInput } from 'react-native';
import { Keyboard } from 'react-native';

import { useSelectContext } from '../../context';

export const useSelectInput = () => {
    const searchInputRef = useRef<TextInput>(null);

    const {
        isOpened,
        disabled,
        multiple,
        placeholderText,
        placeholderTextColor,
        searchPattern,
        searchValue,
        onPressSelectControl,
        selectInputProps,
        dispatch,
        setOptionsListPosition,
        onSelectChangeText,
        styles,
        selectedOption,
    } = useSelectContext();

    const { select: selectStyles } = styles ?? {};
    const { text: textCustomStyles } = selectStyles ?? {};

    useEffect(() => {
        const showSubscription = Keyboard.addListener('keyboardDidShow', async () => {
            await setOptionsListPosition();
        });
        const hideSubscription = Keyboard.addListener('keyboardDidHide', async () => {
            await setOptionsListPosition();
        });
        dispatch({
            type: 'setSearchInputRef',
            payload: searchInputRef,
        });
        return () => {
            showSubscription.remove();
            hideSubscription.remove();
            dispatch({
                type: 'setSearchInputRef',
                payload: null,
            });
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const onChangeText = (payload: string) => {
        if (disabled) {
            return;
        }
        if (!isOpened) {
            dispatch({ type: 'open' });
        }
        dispatch({
            type: 'setSearchValue',
            payload,
        });
        if (searchPattern) {
            dispatch({
                type: 'searchOptions',
                searchPattern,
                payload,
            });
        }
        // callback

        onSelectChangeText?.(payload);
    };

    const resolvePlaceholder = () => {
        if (multiple && Array.isArray(selectedOption) && selectedOption.length > 0) {
            return '  ';
        }
        return placeholderText;
    };

    const resolvedPlaceholder = resolvePlaceholder();

    return {
        disabled,
        multiple,
        placeholderTextColor,
        searchValue,
        onPressSelectControl,
        selectInputProps,
        textCustomStyles,
        searchInputRef,
        resolvedPlaceholder,
        onChangeText,
    };
};
