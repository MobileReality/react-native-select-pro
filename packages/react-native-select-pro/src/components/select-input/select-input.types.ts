import type { StyleProp, TextStyle } from 'react-native';

import type { OptionalToRequired } from '../../helpers';
import type { State } from '../../state';

export type SelectInputProps<T> = OptionalToRequired<
    Pick<State<T>, 'selectedOption'> & {
        textStyle: StyleProp<TextStyle>;
    }
>;
