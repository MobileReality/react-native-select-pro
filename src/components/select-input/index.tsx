import React, { useEffect, useRef } from 'react';
import { Keyboard, StyleSheet, TextInput, TextStyle } from 'react-native';
import type {
    OnPressSelectControlType,
    OnSetPosition,
    SelectProps,
} from '@mobile-reality/react-native-select-pro';

import { COLORS, FONT_SIZE } from '../../constants/styles';
import type { OptionalToRequired } from '../../helpers';
import type { DispatchType, State } from '../../state/types';
import { Action } from '../../state/types';

type SelectInputProps = OptionalToRequired<
    Pick<State, 'isOpened' | 'searchValue'> & { dispatch: DispatchType } & Pick<
            SelectProps,
            'placeholderText' | 'searchPattern' | 'disabled'
        > & { onPressSelectControl: OnPressSelectControlType } & { setPosition: OnSetPosition }
>;

export const SelectInput = ({
    disabled,
    isOpened,
    searchValue,
    searchPattern,
    placeholderText,
    onPressSelectControl,
    dispatch,
    setPosition,
}: SelectInputProps) => {
    const searchInputRef = useRef<TextInput>(null);

    useEffect(() => {
        const showSubscription = Keyboard.addListener('keyboardDidShow', () => {
            setPosition();
        });
        const hideSubscription = Keyboard.addListener('keyboardDidHide', () => {
            setPosition();
        });
        dispatch({
            type: Action.SetSearchInputRef,
            payload: searchInputRef,
        });
        return () => {
            showSubscription.remove();
            hideSubscription.remove();
            dispatch({
                type: Action.SetSearchInputRef,
                payload: null,
            });
        };
    }, []);

    const onChangeText = (payload: string) => {
        if (!disabled) {
            if (!isOpened) {
                dispatch({ type: Action.Open });
            }
            dispatch({
                type: Action.SetSearchValue,
                payload,
            });
            if (searchPattern) {
                dispatch({
                    type: Action.SearchOptions,
                    searchPattern,
                    payload,
                });
            }
        }
    };

    return (
        <TextInput
            accessibilityLabel={'Place text'}
            editable={!disabled}
            onChangeText={onChangeText}
            onPressIn={!disabled ? onPressSelectControl : () => null}
            placeholder={placeholderText}
            ref={searchInputRef}
            style={disabled ? [styles.disabled, styles.text] : styles.text}
            value={searchValue}
        />
    );
};

type Styles = {
    text: TextStyle;
    disabled: TextStyle;
};

const styles = StyleSheet.create<Styles>({
    text: {
        fontSize: FONT_SIZE,
    },
    disabled: {
        backgroundColor: COLORS.DISABLED,
    },
});
