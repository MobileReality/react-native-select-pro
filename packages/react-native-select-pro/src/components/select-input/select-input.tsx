import React from 'react';
import type { TextStyle } from 'react-native';
import { I18nManager, StyleSheet, TextInput } from 'react-native';

import { COLORS, FONT_SIZE } from '../../constants';

import { useSelectInput } from './select-input.hooks';

export const SelectInput = () => {
    const {
        disabled,
        multiple,
        placeholderTextColor,
        searchValue,
        onPressSelectControl,
        selectInputProps,
        textCustomStyles,
        searchInputRef,
        resolvedPlaceholder,
        onChangeText,
    } = useSelectInput();

    return (
        <TextInput
            accessibilityLabel="Place text"
            {...selectInputProps}
            ref={searchInputRef}
            editable={!disabled}
            placeholder={resolvedPlaceholder}
            placeholderTextColor={placeholderTextColor}
            style={
                disabled
                    ? [styles.disabled, styles.text, multiple && styles.marginMultiple]
                    : [styles.text, textCustomStyles]
            }
            textAlign={I18nManager.getConstants().isRTL ? 'right' : 'left'}
            value={searchValue ?? ''}
            onChangeText={onChangeText}
            onPressIn={disabled ? undefined : onPressSelectControl}
        />
    );
};

type Styles = {
    text: TextStyle;
    disabled: TextStyle;
    marginMultiple: TextStyle;
};

const styles = StyleSheet.create<Styles>({
    text: {
        fontSize: FONT_SIZE,
    },
    disabled: {
        backgroundColor: COLORS.DISABLED,
    },
    marginMultiple: {
        marginRight: 6,
    },
});
