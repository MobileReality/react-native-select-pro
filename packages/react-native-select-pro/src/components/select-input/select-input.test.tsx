import React from 'react';

import { SelectInput } from './select-input';

describe('SelectInput', () => {
    it('should generate SelectInput snapshot', () => {
        const selectInput = (
            <SelectInput
                disabled={false}
                isOpened={false}
                searchValue=""
                searchPattern={undefined}
                textInputProps={{}}
                placeholderText={undefined}
                dispatch={() => {}}
                setPosition={() => {}}
                multiSelection={false}
                selectedOption={null}
                placeholderTextColor={undefined}
                textStyle={{}}
                onPressSelectControl={() => {}}
            />
        );
        expect(selectInput).toMatchSnapshot();
    });
});
