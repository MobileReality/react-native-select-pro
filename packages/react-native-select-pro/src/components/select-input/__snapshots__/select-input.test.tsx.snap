// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput should generate SelectInput snapshot 1`] = `
<SelectInput
  disabled={false}
  dispatch={[Function]}
  isOpened={false}
  multiSelection={false}
  onPressSelectControl={[Function]}
  searchValue=""
  selectedOption={null}
  setPosition={[Function]}
  textInputProps={{}}
  textStyle={{}}
/>
`;
